bin/run_docker.py

```diff
@@ -267,8 +268,55 @@ def mount_volumes() -> dict:
     return volumes
 
 
+def monitor_container(
+    container: docker.models.containers.Container,
+    timeout: Union[int, float],
+    poll_interval: Union[int, float],
+    elapsed_time: int = 0,
+):
+    """
+    Recursively monitor the status of a Docker container until it finishes,
+    or the timeout is reached, at which point the container is killed.
+
+    Arguments:
+        container: The Docker container subject to monitor.
+        timeout: Maximum duration to monitor the container (in minutes) before forced shutdown.
+        poll_interval: Time to wait between status checks (in minutes).
+        elapsed_time: The elapsed time already spent monitoring (in minutes).
+    """
+
+    # Check if the elapsed time has reached or exceeded the timeout
+    if elapsed_time >= timeout:
+        print("Timeout reached. Stopping container.")
+        container.stop(timeout=10)
+        return f"Container exceeded execution time limit of {timeout} minutes. Unable to process."
+
+    # Refresh container status
+    container.reload()
+
+    # If the container has exited, stop monitoring
+    if container.status == "exited":
+        return ""
+
+    # Wait before the next check (convert to seconds)
+    time.sleep(poll_interval * 60)
+
+    # Increment the elapsed time
+    elapsed_time += poll_interval
+
+    # Notify the backend when elapsed time is 3/4 of the way to timeout
+    if elapsed_time >= timeout * 0.75:
+        print(f"Time spent monitoring container (minutes): {elapsed_time}")
+        print(f"Container run will shut down after {timeout} minute(s).")
+
+    # Recursively call the function to continue monitoring
+    return monitor_container(container, timeout, poll_interval, elapsed_time)
```

Comment:
Nit: this could be a `while` loop, but the recursion is unavoidable.  Why you would want to do a  `while` loop is so that people can read the code directly and automatically understand the conditions for the code to keep running.

Reply:
Updated: https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/26/commits/802c3f478ff8a3adf1d7082ea2b73a4954581d44

Reply:
Update: My interpretation of this comment led to adding a `while` loop with an explicit conditional (`while container.status != "exited" or elapsed_time < timeout`) instead of using a `while True`. However, implementing the `while` loop this way while also addressing SonarCloud's requirements complicates the code more than making it easier to read, in my opinion. So what I did instead is turn the while loop into an if statement (https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/26/commits/fa2ce5d3bed9bba3a2da673c269e429b100b524b), and that serves the purpose of giving a reader an at-a-glance view of the conditions that need to be met for the recursion to continue.

let me know if this also works for you